async-coap-uri/src/escape/unescape_uri.rs

```diff
@@ -311,16 +406,20 @@ impl<'a> Iterator for UnescapeUri<'a> {
                         // This is a UTF8 byte.
                         utf8_buf[utf8_len] = decoded;
                         utf8_len += 1;
+
                         if utf8_len >= UTF8_CHAR_WIDTH[utf8_buf[0] as usize] as usize {
                             if let Ok(utf8_str) = from_utf8(&utf8_buf[..utf8_len]) {
                                 return Some(utf8_str.chars().next().unwrap());
                             } else {
-                                self.had_error = true;
+                                self.had_error = Some(UnescapeError::InvalidUtf8 {
+                                    buf: utf8_buf,
+                                    len: utf8_len,
+                                });
                                 return Some(REPLACEMENT_CHARACTER);
                             }
                         }
                     } else if utf8_len != 0 {
-                        self.had_error = true;
+                        self.had_error = Some(UnescapeError::InvalidByte(decoded));
```

Comment:
This is another one, where I am not sure what it is supposed to mean/what the conditions were to reach this error?

Reply:
We reach this condition when we have reach a non-UTF8 character when we are in the process of parsing a UTF8 sequence. Take, for example, the emoji `🐙`: it has a URI-encoded sequence of `%f0%9f%90%99`. This check is to make sure we properly error out if we encounter a sequence like `%f0%9f%90%20`. That is an invalid sequence.